core/src/telemetry/metrics.rs

```diff
@@ -592,6 +593,19 @@ pub(super) const TASK_SLOTS_AVAILABLE_NAME: &str = "worker_task_slots_available"
 pub(super) const TASK_SLOTS_USED_NAME: &str = "worker_task_slots_used";
 pub(super) const STICKY_CACHE_SIZE_NAME: &str = "sticky_cache_size";
 
+/// Calls the provided metric function only if the failure is not a benign application failure.
+pub(crate) fn record_failure_metric(
+    failure: &Option<Failure>,
+    metric_fn: impl FnOnce(),
+) {
+    let is_benign = failure
+        .as_ref()
+        .map_or(false, |f| f.is_benign_application_failure());
+    if !is_benign {
+        metric_fn();
+    }
+}
+
```

Comment:
I don't think we really need this function - the closure passing is a bit performative. I think it'd be easier to call this at the callsite, or, if you want to save a few lines, you can make a trait that handles the option stuff an implement it for `Option<Failure>` and that way you can call `should_record_metric()` directly on the `Option<Failure>`